Fixed accidental displacement in :class:`~.Axes` caused by `include_numbers`/`numbers_to_include`

hydrobeam opened the pull request:
<!--
Thanks for your contribution to ManimCommunity!

Before filling in the details, ensure:
- Your local changes are up-to-date with ManimCommunity/manim
  
- The title of your PR gives a descriptive summary to end-users. Some examples:
  - Fixed last animations not running to completion
  - Added gradient support and documentation for SVG files
  Examples of what *NOT* to do:
  - "fixed that styling issue" - not descriptive enough
  - "fixed issue #XYZ" - end-user needs to do further research
-->
## Changelog / Overview
<!-- Optional (Recommended): a detailed overview of the PR for the upcoming
release's changelog entry. Useful for when the PR title isn't enough. 

DO NOT REMOVE THE FOLLOWING CHANGELOG LINES, EVEN IF YOU DON'T USE THEM.-->
<!--changelog-start-->

<!--changelog-end-->

## Motivation
<!-- In what way do your changes improve the library? -->
Originally, adding `include_numbers` to the `axis_config` for the classes in coordinate system would displace the axes because the axes would be centered with the numbers included. This leads to some issues as described in #1662 and #1624.

Here are some images that show the difference:
```py
class E(Scene):
    def construct(self):
        axes = Axes(axis_config={'include_numbers': True}).set_color(GREEN)
        max = Axes().set_color(PINK)
        f_vec = Vector(axes.coords_to_point(np.cos(45 * PI / 180), np.sin(45 * PI / 180)))
        f_vec.set_stroke(width=2.0)

        self.add(max, axes, f_vec)
```
Pre-Changes (the hanging `-7` displaces the `Axes` when centered):
![image](https://user-images.githubusercontent.com/70682032/121462882-47595b00-c97f-11eb-898a-4be5f81d8145.png)

Post-Changes:
![image](https://user-images.githubusercontent.com/70682032/121463138-77086300-c97f-11eb-9d5f-6d265176cbf0.png)

There was also a fix for axis label displacement due to numbers:
Pre-fix: (y_label is shifted to the left)
```py
class o(Scene):
    def construct(self):
        ax = Axes([-2, 8], [-4, 6], axis_config={"include_numbers":True, "decimal_number_config":{"num_decimal_places":2}}, )
        swag = ax.get_axis_labels()
        self.add(ax, swag)
```
![image](https://user-images.githubusercontent.com/70682032/121606140-9b198200-ca1b-11eb-85e7-12aa9191445e.png)

Post-fix:
![image](https://user-images.githubusercontent.com/70682032/121606212-be443180-ca1b-11eb-8f8f-9e54982e7cf5.png)


## Explanation for Changes
<!-- How do your changes improve the library? -->
Instead of generating the numbers using `create_axis` there is now a separate function called `init_numbers` that checks if `"include_numbers"` or `"numbers_to_include"` is in the config for each axis. 
Also fix position of labels when numbers are included.

The refactoring includes:
`self.<x/y/z>_axis_config` now contains the sum of the configs from `self.axis_config` too. `self.x_axis_config` overwrites what is in `self.axis_config`.
The configs are now tied specifically to an axis via an attribute `axis.config`. 
Removed unnecessary parameters from `NumberPlane` and use `kwargs` instead. 

Also added `get_coordinate_labels` for `ThreeDAxes` since this previously only existed for two dimensions.
<!-- For PRs introducing new features, please adjust
the link below to reference the docs of your feature. -->
[Documentation Reference](https://manimce--####.org.readthedocs.build): N/A
<!-- In the link above replace #### with you PR number.
You can also adjust the path to the module / class you worked on. This could look like adding
"/en/####/reference/manim.mobject.geometry.Arc.html" to the link for the Arc class.
Notice that the link will only work once the documentation is build which may take a few minutes.-->


## Testing Status
<!-- Optional (Recommended): your computer specs and what tests you ran with
their results, if any. This section is also intended for other
testing-related comments. -->

## Further Comments
<!-- Optional: any further comments that might be useful for reviewers. -->
Maybe i should split some of these changes up into different PR's for the sake of the changelog 🤔

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings
<!-- Once again, thanks for contributing to ManimCommunity! -->


<!-- Do not modify the lines below. These are for the reviewers of your PR -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

abhi3019 commented:
Hi, the displaced `Axes`  can be fixed with a fairly simpler approach:
https://github.com/ManimCommunity/manim/blob/1e01feb70dccbe493d173ef76046a3d1b851e685/manim/mobject/coordinate_systems.py#L997
The above line can be replaced with 
```py
axes_lines_bounding_box_center = [
    ((axis.x_max + axis.x_min) * axis.unit_size / 2) for axis in self.axes
]
if len(axes_lines_bounding_box_center) == 2:
    axes_lines_bounding_box_center.append(0)
self.shift(-np.array(axes_lines_bounding_box_center))
```
This way we don't have to create [temp_config](https://github.com/ManimCommunity/manim/pull/1664/files#diff-b6144e02b604d43f8149021177941b176f959cb360e061aace916f5ddcc4742aR1052)s and call [init_numbers](https://github.com/ManimCommunity/manim/pull/1664/files#diff-b6144e02b604d43f8149021177941b176f959cb360e061aace916f5ddcc4742aR1009) later.

hydrobeam commented:
That solution is significantly more elegant than the one I came up with! I adjusted it slightly so that it still functions if the range does not contain 0 (as in `x_range = [4, 8]`)

abhi3019 commented:
> I adjusted it slightly so that it still functions if the range does not contain 0 (as in `x_range = [4, 8]`)

Ah, I forgot to compensate for this scenario. Looks like you've already worked it out, but here is a slight modification of my previous solution:
```py
axes_lines_bounding_box_center = [
    ((axis.x_max + axis.x_min) / 2) for axis in self.axes
]
if len(axes_lines_bounding_box_center) == 2:
    axes_lines_bounding_box_center.append(0)

self.shift(-self.c2p(*axes_lines_bounding_box_center))
```
### Example
```py
class AxesTest(Scene):
    def construct(self):
        axes = Axes(x_range=[2, 5], x_length=5, y_range=[1,4], y_length=4, axis_config={'include_numbers': True})
        self.add(axes, Dot(color=YELLOW))
```
Notice how the geometric center of `Axes` coincides with `ORIGIN` which is the desired result
![Example2_ManimCE_v0 7 0](https://user-images.githubusercontent.com/73361366/121639237-dc587300-caa9-11eb-8786-fe6bcc77d7af.png)

behackl reviewed (approved):
Read through the code, looks good to me. I'll leave this open for a while in case someone else wants to take a look.

abhi3019 commented:
There are currently two methods to add coordinate labels - [one for `Axes`](https://github.com/hydrobeam/manim/blob/9819c1b87148417c4d584a8e446eea13b5d38812/manim/mobject/coordinate_systems.py#L1091-L1096) and the [other for `ThreeDAxes`](https://github.com/hydrobeam/manim/blob/9819c1b87148417c4d584a8e446eea13b5d38812/manim/mobject/coordinate_systems.py#L1349-L1355) (introduced in this PR). Why not just have a single `CoordinateSystem.add_coordinate_labels()` method something like this?
```py
def add_coordinate_labels(self, *axes_labels, **kwargs):
    axes_labels = list(axes_labels)
    while len(axes_labels) != self.dimension:
        axes_labels.append(None)
    for axis, labels in zip(self.axes, axes_labels):
        axis.add_numbers(labels, **kwargs)
``` 

Also, currently the `Axes.get_coordinate_labels()` method doesn't just return the coordinate labels, it also adds them
https://github.com/ManimCommunity/manim/blob/3340ceafbd396419ce3408c6aad6a1370d52e5c2/manim/mobject/coordinate_systems.py#L1101
which makes this method redundant https://github.com/ManimCommunity/manim/blob/3340ceafbd396419ce3408c6aad6a1370d52e5c2/manim/mobject/coordinate_systems.py#L1105-L1120

So imo, we can get rid of both `Axes.get_coordinate_labels` and `Axes.add_coordinates` and substitute them with `CoordinateSystem.add_coordinate_labels` which supports `ThreeDAxes` as well.

hydrobeam commented:
That sounds like a good idea 👍I've added that and kept the name of `add_coordinates` since it was used in two other classes (although now that I think about it... `add_numbers` might make more sense). I also added a doc string with an example use-case.

abhi3019 commented on manim/mobject/coordinate_systems.py:
```diff
@@ -225,6 +225,42 @@ def get_axis_labels(
+        """Adds labels to the axes.
+
+        axes_numbers
+            The the numbers to be added to the axes. Use ``None`` to represent an axis without labels.
```
Passing None will add the coordinates by default, depending on the range terms of that axis...
```suggestion
            The numbers to be added to the axes. Use ``None`` to represent an axis with default labels.
```

abhi3019 commented on manim/mobject/coordinate_systems.py:
```diff
@@ -225,6 +225,42 @@ def get_axis_labels(
+        .. code-block:: python
+
+            ax = ThreeDAxes()
+            x_labels = range(1, 11)
+            z_labels = range(2, 12, 2)
```
Since the default x,y,z ranges of `ThreeDAxes` are (-6, 6, 1), (-5, 5, 1) and (-4, 4, 1) respectively, I think it would be better to have these ranges within the axes's ranges. 
```suggestion
            x_labels = range(-4, 5)
            z_labels = range(-4, 4, 2)
```

abhi3019 commented on manim/mobject/coordinate_systems.py:
```diff
@@ -225,6 +225,42 @@ def get_axis_labels(
+            ax = ThreeDAxes()
+            x_labels = range(1, 11)
+            z_labels = range(2, 12, 2)
+            ax.add_coordinates(x_labels, None, z_labels)  # x & z labels
```
This also adds the default y-labels
```suggestion
            ax.add_coordinates(x_labels, None, z_labels)  # default y labels, custom x & z labels
```

abhi3019 commented on manim/mobject/coordinate_systems.py:
```diff
@@ -225,6 +225,42 @@ def get_axis_labels(
+        self.coordinate_labels = VGroup()
+        # if nothing is passed to axes_numbers, produce axes with default labelling
+        if not axes_numbers:
+            axes_numbers = [None for x in self.axes]
```
Since we aren't really using any value from the `self.axes` iterations...
```suggestion
            axes_numbers = [None for _ in range(self.dimension)]
```

abhi3019 reviewed (approved):
LGTM, thanks!
